File: windows/src/GoalMaker.Infrastructure/Auth/LocalOnlyAuthGateway.cs
using GoalMaker.Core.Account;
using GoalMaker.Core.Auth;

namespace GoalMaker.Infrastructure.Auth;

/// <summary>
/// A dev build's <see cref="IAuthGateway"/>: signed in, as soon as it starts, as one fixed owner who lives
/// only on this PC, so trying a change out never waits on a sign-in (docs/sign-in.md). Sign-in is for
/// release builds, and for a dev build started with <c>--sign-in</c>.
/// </summary>
public sealed class LocalOnlyAuthGateway : IAuthGateway
{
    /// <summary>The owner of every row a dev build writes; the same one the phone's dev build uses.</summary>
    public const string OwnerId = "00000000-0000-4000-8000-00000000d0e0";

    public AuthSession Session { get; private set; } = new AuthSession.Loading();

    public event EventHandler<AuthSession>? SessionChanged;

    public Task InitializeAsync(CancellationToken cancellationToken)
    {
        Session = new AuthSession.SignedIn(OwnerId, string.Empty);
        SessionChanged?.Invoke(this, Session);
        return Task.CompletedTask;
    }

    public Task<AuthResult> SendCodeAsync(EmailAddress email, CancellationToken cancellationToken) =>
        Task.FromResult<AuthResult>(new AuthResult.Success());

    public Task<AuthResult> VerifyCodeAsync(EmailAddress email, SignInCode code, CancellationToken cancellationToken) =>
        Task.FromResult<AuthResult>(new AuthResult.Success());

    public Task SignOutAsync() => Task.CompletedTask;
}

File: android/app/src/main/kotlin/com/goalmaker/app/data/auth/LocalOnlyAuthGateway.kt
package com.goalmaker.app.data.auth

import com.goalmaker.app.application.auth.AuthGateway
import com.goalmaker.app.application.auth.AuthResult
import com.goalmaker.app.application.auth.AuthSession
import com.goalmaker.app.domain.account.EmailAddress
import com.goalmaker.app.domain.account.SignInCode
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow

/**
 * A dev build's [AuthGateway]: always signed in as one fixed owner who lives only on this phone, so
 * trying a change out never waits on a sign-in (docs/sign-in.md). Sign-in is for release builds.
 */
class LocalOnlyAuthGateway : AuthGateway {
    override val session: StateFlow<AuthSession> = MutableStateFlow(AuthSession.SignedIn(OWNER_ID, ""))

    override suspend fun sendCode(email: EmailAddress): AuthResult = AuthResult.Success

    override suspend fun verifyCode(email: EmailAddress, code: SignInCode): AuthResult = AuthResult.Success

    override suspend fun signOut() = Unit

    companion object {
        /** The owner of every row a dev build writes. */
        const val OWNER_ID = "00000000-0000-4000-8000-00000000d0e0"
    }
}
